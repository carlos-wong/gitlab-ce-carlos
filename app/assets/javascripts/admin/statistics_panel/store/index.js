import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import mutations from './mutations';
import state from './state';

Vue.use(Vuex);

export default () =>
  new Vuex.Store({
    actions,
    getters,
    mutations,
    state: state(),
  });
