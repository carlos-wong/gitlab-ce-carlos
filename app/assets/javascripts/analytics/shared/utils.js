import dateFormat from 'dateformat';
import { hideFlash } from '~/flash';
import { slugify } from '~/lib/utils/text_utility';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import { dateFormats } from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  /* eslint-disable camelcase */
  const {
    source_branch_name = null,
    target_branch_name = null,
    author_username = null,
    milestone_title = null,
    assignee_username = [],
    label_name = [],
  } = urlQueryToFilter(url);
  /* eslint-enable camelcase */

  return {
    selectedSourceBranch: source_branch_name,
    selectedTargetBranch: target_branch_name,
    selectedAuthor: author_username,
    selectedMilestone: milestone_title,
    selectedAssigneeList: assignee_username,
    selectedLabelList: label_name,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

/**
 * Prepares metric data to be rendered in the metric_card component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @param {Object} popoverContent - Key value pair of data to display in the popover
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_card
 */
export const prepareTimeMetricsData = (data = [], popoverContent = {}) =>
  data.map(({ title: label, identifier, ...rest }) => {
    const metricIdentifier = identifier || slugify(label);
    return {
      ...rest,
      label,
      identifier: metricIdentifier,
      description: popoverContent[metricIdentifier]?.description || '',
    };
  });
