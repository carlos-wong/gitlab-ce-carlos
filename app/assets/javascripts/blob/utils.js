/* global ace */
import Editor from '~/editor/editor_lite';

export function initEditorLite({ el, blobPath, blobContent }) {
  if (!el) {
    throw new Error(`"el" parameter is required to initialize Editor`);
  }
  let editor;

  if (window?.gon?.features?.monacoSnippets) {
    editor = new Editor();
    editor.createInstance({
      el,
      blobPath,
      blobContent,
    });
  } else {
    editor = ace.edit(el);
  }

  return editor;
}

export default () => ({});
