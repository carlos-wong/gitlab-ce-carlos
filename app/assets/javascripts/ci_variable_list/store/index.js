import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import mutations from './mutations';
import state from './state';

Vue.use(Vuex);

export default (initialState = {}) =>
  new Vuex.Store({
    actions,
    mutations,
    state: {
      ...state(),
      ...initialState,
    },
  });
