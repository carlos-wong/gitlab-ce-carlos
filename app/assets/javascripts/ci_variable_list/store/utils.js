import { cloneDeep } from 'lodash';
import { displayText, types } from '../constants';

const variableTypeHandler = type =>
  type === displayText.variableText ? types.variableType : types.fileType;

export const prepareDataForDisplay = variables => {
  const variablesToDisplay = [];
  variables.forEach(variable => {
    const variableCopy = variable;
    if (variableCopy.variable_type === types.variableType) {
      variableCopy.variable_type = displayText.variableText;
    } else {
      variableCopy.variable_type = displayText.fileText;
    }
    variableCopy.secret_value = variableCopy.value;

    if (variableCopy.environment_scope === types.allEnvironmentsType) {
      variableCopy.environment_scope = displayText.allEnvironmentsText;
    }
    variablesToDisplay.push(variableCopy);
  });
  return variablesToDisplay;
};

export const prepareDataForApi = (variable, destroy = false) => {
  const variableCopy = cloneDeep(variable);
  variableCopy.protected = variableCopy.protected.toString();
  variableCopy.masked = variableCopy.masked.toString();
  variableCopy.variable_type = variableTypeHandler(variableCopy.variable_type);
  if (variableCopy.environment_scope === displayText.allEnvironmentsText) {
    variableCopy.environment_scope = types.allEnvironmentsType;
  }

  if (destroy) {
    // eslint-disable-next-line
    variableCopy._destroy = destroy;
  }

  return variableCopy;
};

export const prepareEnvironments = environments => environments.map(e => e.name);
