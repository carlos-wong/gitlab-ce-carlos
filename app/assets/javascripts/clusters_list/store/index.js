import Vue from 'vue';
import Vuex from 'vuex';
import state from './state';
import mutations from './mutations';
import * as actions from './actions';

Vue.use(Vuex);

export const createStore = initialState =>
  new Vuex.Store({
    actions,
    mutations,
    state: state(initialState),
  });

export default createStore;
