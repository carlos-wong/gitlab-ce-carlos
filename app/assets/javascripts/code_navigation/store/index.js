import Vuex from 'vuex';
import createState from './state';
import actions from './actions';
import mutations from './mutations';

export default new Vuex.Store({
  actions,
  mutations,
  state: createState(),
});
