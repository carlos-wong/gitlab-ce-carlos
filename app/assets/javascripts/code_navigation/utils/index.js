export const cachedData = new Map();

export const getCurrentHoverElement = () => cachedData.get('current');
export const setCurrentHoverElement = el => cachedData.set('current', el);

export const addInteractionClass = d => {
  let charCount = 0;
  const line = document.getElementById(`LC${d.start_line + 1}`);
  const el = [...line.childNodes].find(({ textContent }) => {
    if (charCount === d.start_char) return true;
    charCount += textContent.length;
    return false;
  });

  if (el) {
    el.setAttribute('data-char-index', d.start_char);
    el.setAttribute('data-line-index', d.start_line);
    el.classList.add('cursor-pointer', 'code-navigation', 'js-code-navigation');
  }
};
