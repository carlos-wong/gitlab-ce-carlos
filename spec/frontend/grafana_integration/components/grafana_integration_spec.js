import { mount, shallowMount } from '@vue/test-utils';
import { GlButton } from '@gitlab/ui';
import GrafanaIntegration from '~/grafana_integration/components/grafana_integration.vue';
import { createStore } from '~/grafana_integration/store';
import axios from '~/lib/utils/axios_utils';
import { refreshCurrentPage } from '~/lib/utils/url_utility';
import createFlash from '~/flash';
import { TEST_HOST } from 'helpers/test_constants';

jest.mock('~/lib/utils/url_utility');
jest.mock('~/flash');

describe('grafana integration component', () => {
  let wrapper;
  let store;
  const operationsSettingsEndpoint = `${TEST_HOST}/mock/ops/settings/endpoint`;
  const grafanaIntegrationUrl = `${TEST_HOST}`;
  const grafanaIntegrationToken = 'someToken';

  beforeEach(() => {
    store = createStore({
      operationsSettingsEndpoint,
      grafanaIntegrationUrl,
      grafanaIntegrationToken,
    });
  });

  afterEach(() => {
    if (wrapper.destroy) {
      wrapper.destroy();
      createFlash.mockReset();
      refreshCurrentPage.mockReset();
    }
  });

  describe('default state', () => {
    it('to match the default snapshot', () => {
      wrapper = shallowMount(GrafanaIntegration, { store });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  it('renders header text', () => {
    wrapper = shallowMount(GrafanaIntegration, { store });

    expect(wrapper.find('.js-section-header').text()).toBe('Grafana Authentication');
  });

  describe('expand/collapse button', () => {
    it('renders as an expand button by default', () => {
      wrapper = shallowMount(GrafanaIntegration, { store });

      const button = wrapper.find(GlButton);

      expect(button.text()).toBe('Expand');
    });
  });

  describe('sub-header', () => {
    it('renders descriptive text', () => {
      wrapper = shallowMount(GrafanaIntegration, { store });

      expect(wrapper.find('.js-section-sub-header').text()).toContain(
        'Embed Grafana charts in GitLab issues.',
      );
    });
  });

  describe('form', () => {
    beforeEach(() => {
      jest.spyOn(axios, 'patch').mockImplementation();
    });

    afterEach(() => {
      axios.patch.mockReset();
    });

    describe('submit button', () => {
      const findSubmitButton = () => wrapper.find('.settings-content form').find(GlButton);

      const endpointRequest = [
        operationsSettingsEndpoint,
        {
          project: {
            grafana_integration_attributes: {
              grafana_url: grafanaIntegrationUrl,
              token: grafanaIntegrationToken,
              enabled: false,
            },
          },
        },
      ];

      it('submits form on click', () => {
        wrapper = mount(GrafanaIntegration, { store });
        axios.patch.mockResolvedValue();

        findSubmitButton(wrapper).trigger('click');

        expect(axios.patch).toHaveBeenCalledWith(...endpointRequest);
        return wrapper.vm.$nextTick().then(() => expect(refreshCurrentPage).toHaveBeenCalled());
      });

      it('creates flash banner on error', () => {
        const message = 'mockErrorMessage';
        wrapper = mount(GrafanaIntegration, { store });
        axios.patch.mockRejectedValue({ response: { data: { message } } });

        findSubmitButton().trigger('click');

        expect(axios.patch).toHaveBeenCalledWith(...endpointRequest);
        return wrapper.vm
          .$nextTick()
          .then(jest.runAllTicks)
          .then(() =>
            expect(createFlash).toHaveBeenCalledWith(
              `There was an error saving your changes. ${message}`,
              'alert',
            ),
          );
      });
    });
  });
});
