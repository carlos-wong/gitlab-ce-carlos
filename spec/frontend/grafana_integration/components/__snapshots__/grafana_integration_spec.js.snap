// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grafana integration component default state to match the default snapshot 1`] = `
<section
  class="settings no-animate js-grafana-integration"
  id="grafana"
>
  <div
    class="settings-header"
  >
    <h4
      class="js-section-header"
    >
      
      Grafana Authentication
    
    </h4>
     
    <gl-button-stub
      class="js-settings-toggle"
      size="md"
      variant="secondary"
    >
      Expand
    </gl-button-stub>
     
    <p
      class="js-section-sub-header"
    >
      
      Embed Grafana charts in GitLab issues.
    
    </p>
  </div>
   
  <div
    class="settings-content"
  >
    <form>
      <gl-form-checkbox-stub
        class="mb-4"
        id="grafana-integration-enabled"
      >
        
        Active
      
      </gl-form-checkbox-stub>
       
      <gl-form-group-stub
        description="Enter the base URL of the Grafana instance."
        label="Grafana URL"
        label-for="grafana-url"
      >
        <gl-form-input-stub
          id="grafana-url"
          placeholder="https://my-url.grafana.net/"
          value="http://test.host"
        />
      </gl-form-group-stub>
       
      <gl-form-group-stub
        label="API Token"
        label-for="grafana-token"
      >
        <gl-form-input-stub
          id="grafana-token"
          value="someToken"
        />
         
        <p
          class="form-text text-muted"
        >
          
          Enter the Grafana API Token.
          
          <a
            href="https://grafana.com/docs/http_api/auth/#create-api-token"
            rel="noopener noreferrer"
            target="_blank"
          >
            
            More information
            
            <icon-stub
              class="vertical-align-middle"
              name="external-link"
              size="16"
            />
          </a>
        </p>
      </gl-form-group-stub>
       
      <gl-button-stub
        size="md"
        variant="success"
      >
        
        Save Changes
      
      </gl-button-stub>
    </form>
  </div>
</section>
`;
